[Jin Mei Xuan] iP

meixuanjin opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/meixuanjin/ip/releases/tag/A-Jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

*And it is **FREE**!*

Features:
- [x] Week 1 Features
- [x] Week 2 Features
- [x] Week 3 Features
- [ ] Week 4 Features (coming soon!)

## Running the program
You can run the `main` method in Duke.java. It should look like this (sorry it's long!):

```java
public static void main(String[] args) throws IOException {
    Scanner sc = new Scanner(System.in);

    Ui.sayHello();
    String task = sc.nextLine();
    Storage.readFromFile();

    while(!task.equals("bye")) {
        try {
            if (task.equals("list")) {
                TaskList.printList();
            } else if (task.startsWith("done")) {
                TaskList.complete(task);
                Storage.overwrite();
            } else if (task.startsWith("find")) {
                String[] input = task.split("find ", 2);
                TaskList.find(input[1]);
            }else if (task.startsWith("deadline") || task.startsWith("event") || task.startsWith("todo")) {
                Storage.writeToFile(task);
            } else if (task.startsWith("delete")) {
                String numString = task.replaceAll("[^0-9]", "");
                int num = Integer.parseInt(numString);
                TaskList.deleteTask(num);
                Storage.overwrite();
            } else {
                throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
            }
        } catch (DukeException e) {
            System.out.println(e.getMessage());
        }
        task = sc.nextLine();
    }
    System.out.println(LINE);
    System.out.println("Bye. Hope to see you again soon!");
    System.out.println(LINE);
    sc.close();
}
```

Commit:
Add Gradle support

meixuanjin committed:
Level-1

meixuanjin committed:
Level-2

meixuanjin committed:
Level-3

meixuanjin committed:
Refactor code

meixuanjin committed:
Level-4 Todos, Events, Deadlines without I/O redirection technique

meixuanjin committed:
A-TextUiTesting: Automated Text UI Testing

meixuanjin committed:
Level 5: Handle errors

meixuanjin committed:
Level 6: Delete

meixuanjin committed:
Create class files

meixuanjin committed:
Create class files

meixuanjin committed:
Level 7: Save

meixuanjin committed:
Create new class file

meixuanjin committed:
Level 8: Dates and Times

meixuanjin committed:
Merge branches and fix conflicts

meixuanjin committed:
Add Ui Class

meixuanjin committed:
Add Parser Class

meixuanjin committed:
Organise into Packages

meixuanjin committed:
Add JUnit Tests

meixuanjin committed:
Add JavaDocs

meixuanjin committed:
Tweak code to comply with coding standard

meixuanjin committed:
Merge JavaDoc and CodingStandards, fix clashes

meixuanjin committed:
Add Level-9: Find

sebbycake commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+    static void writeToFile(String task) throws IOException {
+        try {
+            Parser.parseTask(task);
+            if (task.startsWith("deadline") ||
+                    (task.startsWith("event") && task.contains("/at ")) ||
+                    (task.startsWith("todo"))) {
```
Perhaps you can break down your Boolean checks into variables to avoid having a complicated expression!

meixuanjin replied:
That makes sense, thank you!

sebbycake commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+ * Represents the file containing the list of tasks.
+ */
+public class Storage {
+    private static final String text = "./data/duke.txt";
```
Since the `text` variable represents a constant file path, perhaps you can rename it to `FILE_PATH` to adhere to Java coding standard and makes it more descriptive.

sebbycake commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+ * Represents a Deadline task.
+ */
+public class Deadline extends Tasks {
+    protected static final DateTimeFormatter formatted = DateTimeFormatter.ofPattern("MMM d yyyy HHmm");
```
Again, can rename the constant to `FORMATTED`

sebbycake commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+                curr[1] = line.substring(4,5);
+                curr[2] = line.substring(7);
+                switch (curr[1]) {
+                    case "T" -> {
+                        TaskList.addSpecificTask("todo " + curr[2]);
+                        if (curr[1].equals("X")) {
+                            TaskList.getLast().markAsDone();
+                        }
+                    }
+                    case "E" -> {
+                        TaskList.addSpecificTask("event " + curr[2]);
+                        if (curr[1].equals("X")) {
+                            TaskList.getLast().markAsDone();
+                        }
+                    }
+                    case "D" -> {
+                        TaskList.addSpecificTask("deadline " + curr[2]);
+                        if (curr[1].equals("X")) {
+                            TaskList.getLast().markAsDone();
+                        }
+                    }
+                    default -> {
+                    }
```
`case` statement do not need to be indented!

sebbycake reviewed (commented):
Hi Mei Xuan, you generally have good code quality! No sights of long methods or deeply nested if-else statements. Method names are mostly self-explanatory with sufficient level of detail.
I have also added some comments for Java coding standard.

meixuanjin committed:
Merge branch 'add-gradle-support'

meixuanjin committed:
Use Gradle, tweak tests

meixuanjin committed:
Use checkStyle to detect coding style violations

meixuanjin committed:
Fix CheckStyle

meixuanjin committed:
Add GUI to Duke

pss-coder commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+    static void writeToFile(String task) throws IOException {
+        try {
+            Parser.parseTask(task);
+            if (task.startsWith("deadline")
+                    || (task.startsWith("event") && task.contains("/at "))
+                    || (task.startsWith("todo"))) {
```
Perhaps you could further unpack this into variables instead to simplify the code readablity and logic ?
for example: 
boolean isDeadline = task.startsWith("deadline");
...

pss-coder commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+            writer = new BufferedWriter(fWriter);
+
+            //reading data
+            while ((line = reader.readLine()) != null) {
+                String[] curr = new String[4];
+                curr[0] = line.substring(1, 2);
+                curr[1] = line.substring(4, 5);
+                curr[2] = line.substring(7);
+                switch (curr[1]) {
+                case "T":
+                    TaskList.addSpecificTask("todo " + curr[2]);
+                    if (curr[1].equals("X")) {
+                        TaskList.getLast().markAsDone();
+                    }
+                case "E":
+                    TaskList.addSpecificTask("event " + curr[2]);
+                    if (curr[1].equals("X")) {
+                        TaskList.getLast().markAsDone();
+                    }
+                case "D":
+                    TaskList.addSpecificTask("deadline " + curr[2]);
+                    if (curr[1].equals("X")) {
+                        TaskList.getLast().markAsDone();
+                    }
+                default: break;
+                }
+                reader.close();
+            }
```
Neat implementation. Given the long lines of code you have, perhaps to improve the code quality you could unpack the lines to another method instead, for example parseData() ?

pss-coder commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+     * Rewrites the duke.txt file with the new list of tasks.
+     */
+    public static void overwrite() throws IOException {
+        FileWriter fWriter = new FileWriter(text, false);
+        writer = new BufferedWriter(fWriter);
+        writer.write(TaskList.overwrite());
+        writer.close();
+        FileWriter newWriter = new FileWriter(text, true);
+        writer = new BufferedWriter(newWriter);
```
Like the neat code.
Perhaps some comments could be included to better understand what is happening or perhaps partition them into different methds ?

pss-coder commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        System.out.println("Nice! I've marked this task as done:");
+        String numString = task.replaceAll("[^0-9]", "");
+        int num = Integer.parseInt(numString);
+        for (int counter = 0; counter < taskString.size(); counter++) {
+            if (counter == num - 1) {
+                taskList.get(num - 1).markAsDone();
+            }
```
Could the code here be better explained, may be adding some code ?

pss-coder commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+    public static void complete(String task) {
+        System.out.println(LINE);
+        System.out.println("Nice! I've marked this task as done:");
+        String numString = task.replaceAll("[^0-9]", "");
```
Cool use of regex.
Could you comment what the regex is representing ?

pss-coder commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,56 @@
+        String task = sc.nextLine();
+        Storage.readFromFile();
+
+        while (!task.equals("bye")) {
+            try {
+                if (task.equals("list")) {
+                    TaskList.printList();
+                } else if (task.startsWith("done")) {
+                    TaskList.complete(task);
+                    Storage.overwrite();
+                } else if (task.startsWith("find")) {
+                    String[] input = task.split("find ", 2);
+                    TaskList.find(input[1]);
+                } else if (task.startsWith("deadline") || task.startsWith("event") || task.startsWith("todo")) {
+                    Storage.writeToFile(task);
+                } else if (task.startsWith("delete")) {
+                    String numString = task.replaceAll("[^0-9]", "");
+                    int num = Integer.parseInt(numString);
+                    TaskList.deleteTask(num);
+                    Storage.overwrite();
+                } else {
+                    throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
```
Great implementation. 
Perhaps to further improve code quality, you could consider to seperate the implementation into a different method, for example parseCommand() ?

pss-coder reviewed (commented):
Hi Mei Xuan, as part of my tutorial for week 5, I am allocated to PR your code for any code quality issues. Besides the few I have commented, all looks good.

Perhaps a suggestion you could consider, given you have alot of String constants in your code, you could consider creating another java file that stores all your String constants for ease of access and modification. Hope this helps :)


Just a couple of nits to fix but. LGTM 👍🏽 😄

meixuanjin committed:
Fix Storage

meixuanjin committed:
Use assertions in Parser and Storage

Use assert feature to document important assumptions
that should hold at various points in the code.

meixuanjin committed:
Use assertions in Parser and Storage

Use assert feature to document important assumptions
that should hold at various points in the code.

meixuanjin committed:
Improve code quality

Refactor code to improve the code quality
where necessary.

meixuanjin committed:
Merge pull request #2 from meixuanjin/branch-A-Assertions

Use assertions in Parser and Storage

meixuanjin committed:
Merge branch 'master' into branch-A-CodeQuality

meixuanjin committed:
Merge pull request #3 from meixuanjin/branch-A-CodeQuality

Improve Code Quality

meixuanjin committed:
Add C-Update to edit items

meixuanjin committed:
Improve GUI

meixuanjin committed:
Add UI Screenshot

meixuanjin committed:
Update README.md

meixuanjin committed:
Set theme jekyll-theme-minimal

meixuanjin committed:
Release v0.2

meixuanjin committed:
Merge remote-tracking branch 'origin/master'

meixuanjin committed:
Clean up code

meixuanjin committed:
Final changes
